### PR TITLE
Bsongis/pxx2

### PR DESCRIPTION
**Pull request by bsongis**
For comments only, not ready yet

### CODE REVIEW

File: radio/src/pulses/pulses_arm.h [hunk 1]
```diff
@@ -86,7 +74,7 @@ PACK(struct Dsm2TimerPulsesData {
 });
 #endif
 
-#if defined(INTMODULE_USART) || defined(EXTMODULE_USART)
+#if defined(INTMODULE_USART) || defined(EXTMODULE_USART) || defined(PXX2)
```
Comment (bsongis) [inline radio/src/pulses/pulses_arm.h]: TODO: this file should not be included when building the bootloader, this test shouldn't be needed

File: radio/src/pulses/pulses_arm.h [hunk 2]
```diff
@@ -125,19 +113,29 @@ PACK(struct PxxTimerPulsesData {
...
+
 #define CROSSFIRE_FRAME_MAXLEN         64
 PACK(struct CrossfirePulsesData {
   uint8_t pulses[CROSSFIRE_FRAME_MAXLEN];
 });
 
 union ModulePulsesData {
-#if defined(INTMODULE_USART) || defined(EXTMODULE_USART)
-  PxxUartPulsesData pxx_uart;
+#if defined(INTMODULE_USART) || defined(EXTMODULE_USART) && !defined(BOOT)
```
Comment (bsongis) [inline radio/src/pulses/pulses_arm.h]: Same comment here

**Comment by bsongis** [comment]
TODO: check if the binary size increased with all those templates

File: radio/src/pulses/pulses_common.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+#if defined(PCBX12S) && PCBREV < 13
+  #define pulse_duration_t             uint32_t
+  #define trainer_pulse_duration_t     uint16_t
+#else
+  #define pulse_duration_t             uint16_t
+  #define trainer_pulse_duration_t     uint16_t
+#endif
+
+template <class T, int SIZE>
```
Comment (bsongis) [inline radio/src/pulses/pulses_common.h:34]: TODO: Check if this class exists somewhere else

File: radio/src/pulses/pxx.h [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    void addTail()
+    {
+      // rest min value is 18000 - 200 * 48 = 8400 (4.2ms)
+      *(ptr - 1) += rest;
+    }
+};
+
+template <class BitTransport>
+class StandardPxxTransport: public PxxCrcMixin, public BitTransport {
```
Comment (bsongis) [inline radio/src/pulses/pxx.h]: TODO: Move BitTransport before CrcMixin to prevent data alignment issues

File: radio/src/pulses/pxx1.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _PULSES_PXX1_H_
+#define _PULSES_PXX1_H_
+
+#include "pxx.h"
+
+template <class PxxTransport>
```
Comment (bsongis) [inline radio/src/pulses/pxx1.h]: Pxx1Transport?

File: radio/src/pulses/pulses_common.h (hunk 3 above)
Comment (3djc) [reply radio/src/pulses/pulses_common.h:34]: Xlite
 ------------- 
2.3
Section Details:
section              size(B)         addr   loc
.text                 368952   0x08000000   Flash
.init_array               36   0x0805a138   Flash
.fini_array                4   0x0805a15c   Flash
.data (added)            384   0x0805a160   Flash
.data                    384   0x20000000   SRAM1
.bss                   50068   0x20000180   SRAM1
._user_heap_stack       1024   0x2000c514   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   360.72   70.45%     512	 151.28
  SRAM1:    50.27   44.88%     112	  61.73
  SRAM2:     0.00    0.00%      16	  16.00
 
PXX2
Does not compile

X7
 ------------- 
2.3
Section Details:
section              size(B)         addr   loc
.text                 370000   0x08000000   Flash
.init_array               36   0x0805a550   Flash
.fini_array                4   0x0805a574   Flash
.data (added)            388   0x0805a578   Flash
.data                    388   0x20000000   SRAM1
.bss                   50156   0x20000188   SRAM1
._user_heap_stack       1024   0x2000c574   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   361.75   70.65%     512	 150.25
  SRAM1:    50.36   44.96%     112	  61.64
  SRAM2:     0.00    0.00%      16	  16.00
  
PXX2
Section Details:
section              size(B)         addr   loc
.text                 370960   0x08000000   Flash
.init_array               36   0x0805a910   Flash
.fini_array                4   0x0805a934   Flash
.data (added)            388   0x0805a938   Flash
.data                    388   0x20000000   SRAM1
.bss                   50156   0x20000188   SRAM1
._user_heap_stack       1024   0x2000c574   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   362.68   70.84%     512	 149.32
  SRAM1:    50.36   44.96%     112	  61.64
  SRAM2:     0.00    0.00%      16	  16.00

X9DP
 ------------- 
2.3
Section Details:
section              size(B)         addr   loc
.text                 381080   0x08000000   Flash
.init_array               36   0x0805d098   Flash
.fini_array                4   0x0805d0bc   Flash
.data (added)            388   0x0805d0c0   Flash
.data                    388   0x20000000   SRAM1
.bss                   58772   0x20000188   SRAM1
._user_heap_stack       1024   0x2000e71c   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   372.57   72.77%     512	 139.43
  SRAM1:    58.77   52.48%     112	  53.23
  SRAM2:     0.00    0.00%      16	  16.00

PXX2 
Section Details:
section              size(B)         addr   loc
.text                 380992   0x08000000   Flash
.init_array               36   0x0805d040   Flash
.fini_array                4   0x0805d064   Flash
.data (added)            388   0x0805d068   Flash
.data                    388   0x20000000   SRAM1
.bss                   58764   0x20000188   SRAM1
._user_heap_stack       1024   0x2000e714   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   372.48   72.75%     512	 139.52
  SRAM1:    58.77   52.47%     112	  53.23
  SRAM2:     0.00    0.00%      16	  16.00
  
  X9E
 -------------   
 2.3
 Section Details:
section              size(B)         addr   loc
.text                 409544   0x08000000   Flash
.init_array               40   0x08063fc8   Flash
.fini_array                4   0x08063ff0   Flash
.data (added)            408   0x08063ff4   Flash
.data                    408   0x10000000   CCRAM
.bss                   44384   0x10000198   CCRAM
.ram                   19584   0x20000000   SRAM1
._user_heap_stack       1024   0x20004c80   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   400.39   78.20%     512	 111.61
  CCRAM:    43.74   68.35%      64	  20.26
  SRAM1:    20.12   17.97%     112	  91.88
  SRAM2:     0.00    0.00%      16	  16.00
  
 PXX2 
 Section Details:
section              size(B)         addr   loc
.text                 409456   0x08000000   Flash
.init_array               40   0x08063f70   Flash
.fini_array                4   0x08063f98   Flash
.data (added)            408   0x08063f9c   Flash
.data                    408   0x10000000   CCRAM
.bss                   44384   0x10000198   CCRAM
.ram                   19584   0x20000000   SRAM1
._user_heap_stack       1024   0x20004c80   SRAM1

Totals         KB    usage  of ttl	KB free
  Flash:   400.30   78.18%     512	 111.70
  CCRAM:    43.74   68.35%      64	  20.26
  SRAM1:    20.12   17.97%     112	  91.88 
  
 X12S
 ------------- 
 2.3 
 Section Details:
section              size(B)         addr   loc
.text                1021184   0x08000000   Flash
.init_array              116   0x080f9500   Flash
.fini_array                4   0x080f9574   Flash
.data (added)            772   0x080f9578   Flash
.data                    772   0x10000000   CCRAM
.bss                   49704   0x10000308   CCRAM
.noinit                    4   0x1000c530   CCRAM
.ram                   16748   0x20000000   SRAM1
._user_heap_stack       1024   0x2000416c   SRAM1
.sdram                783360   0xd0000000   SDRAM2

Totals         KB    usage  of ttl	KB free
  Flash:   998.12   48.74%    2048	1049.88
  CCRAM:    49.30   77.03%      64	  14.70
  SRAM1:    17.36   15.50%     112	  94.64
  SRAM2:     0.00    0.00%      16	  16.00
  SRAM3:     0.00    0.00%      64	  64.00
 SDRAM2:   765.00    9.34%    8192	7427.00
 
 PXX2 
Does not compile

**Comment by bsongis** [comment]
Tested on XLITE

**Pull request merged**